### PR TITLE
Add different roles and update add command

### LINKED ISSUE #85: Modify add command to specify type of person to add

### PR DESCRIPTION
**Pull request by zzzlou**
Fixes #85

### CODE REVIEW

File: src/main/java/seedu/address/model/person/PersonFactory.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+     * its hours value is used as the default. Otherwise, {@code editPersonDescriptor} must contain a non-null hours
+     * value.
+     *
+     * @param descriptor The descriptor containing updated fields.
+     * @param person     The original person whose role-specific field may be used as a default.
+     * @return The {@code Hours} for the new or updated {@code Volunteer}.
+     * @throws CommandException If the hours field is missing and no default is available.
+     */
+    private static Hours retrieveVolunteerHours(EditCommand.EditPersonDescriptor descriptor, Person person)
+            throws CommandException {
```
Comment (Justin-Yeo) [inline src/main/java/seedu/address/model/person/PersonFactory.java:148]: Would an exception here be classified as CommandException or would it be another exception like IllegalArgumentException?

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java [hunk 2]
```diff
@@ -103,7 +135,28 @@ public Person toModelType() throws IllegalValueException {
...
+            return new Donor(modelName, modelPhone, modelEmail, modelAddress, modelTags, modelDonatedAmount);
+        } else if (partnershipEndDate != null) {
+            if (!Date.isValidDate(partnershipEndDate)) {
+                throw new IllegalValueException(Date.MESSAGE_CONSTRAINTS);
+            }
+            final Date modelPartnershipEndDate = new Date(partnershipEndDate);
+            return new Partner(modelName, modelPhone, modelEmail, modelAddress, modelTags, modelPartnershipEndDate);
+        } else {
+            throw new IllegalValueException("Person type information missing. Unable to determine role.");
+        }
```
Comment (Justin-Yeo) [inline src/main/java/seedu/address/storage/JsonAdaptedPerson.java:159]: Are we able to SLAP this more?

**Review by Justin-Yeo — CHANGES_REQUESTED** [review]
Overall looks good! Just some comments and clarifications 👍

File: src/main/java/seedu/address/model/person/Date.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+     * @return True if the test string is a valid date in the specified format.
+     */
+    public static boolean isValidDate(String test) {
+        try {
+            LocalDate.parse(test, FORMATTER);
+            return true;
+        } catch (DateTimeParseException e) {
+            return false;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (chongsun2002) [reply src/main/java/seedu/address/model/person/Date.java:50]: Lets create a separate method to do this, isValidDate should just return the validity of the string format.

File: src/main/java/seedu/address/model/person/PersonFactory.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+     * @param hours The hours contributed, applicable for {@code Volunteer}; otherwise, {@code null}.
+     * @param donatedAmount The donated amount, applicable for {@code Donor}; otherwise, {@code null}.
+     * @param partnershipEndDate The partnership end date, applicable for {@code Partner}; otherwise, {@code null}.
+     * @return A new instance of the appropriate {@code Person} subclass based on the provided fields.
+     * @throws IllegalArgumentException If none of the subclass-specific fields are provided, or if
+     *         more than one subclass-specific field is non-null.
+     */
+    public static Person createPerson(Name name, Phone phone, Email email, Address address, Set<Tag> tags,
+                                      Hours hours, DonatedAmount donatedAmount, Date partnershipEndDate) {
+        if (hours != null) {
```
Comment (chongsun2002) [inline src/main/java/seedu/address/model/person/PersonFactory.java:48]: If volunteers' hours is null, should just set it to 0

**Pull request merged**